File: dev/OpenPipeTest.js
// WinCC Unified Open Pipe — minimal Node.js test program.
//
// Uses net.createConnection which on Windows opens named pipes via the proper
// CreateNamedPipe client API (not as a plain file), giving correct duplex semantics.
//
// Sequence: probe → BrowseTags (paged) → SubscribeTag → watch live notifications
// Each NotifySubscribeTag line shows how many tags changed vs. were suppressed by hasChanged=false.
//
// Requirements:
//   - Node.js 14+
//   - WinCC Unified Runtime running on this host
//   - The user running this command is a member of group "SIMATIC HMI"
//
// Run from the repo root:
//   node dev\OpenPipeTest.js                  (default filter "*", probe first)
//   node dev\OpenPipeTest.js "HMI_Tag_*"      (custom filter)
//   node dev\OpenPipeTest.js skip "HMI_Tag_*" (skip probe, go straight to BrowseTags)

'use strict';

const net = require('net');
const readline = require('readline');

const args = process.argv.slice(2);
let skipProbe = false;
let filter = '*';
if (args[0] && args[0].toLowerCase() === 'skip') {
    skipProbe = true;
    filter = args[1] || '*';
} else if (args[0]) {
    filter = args[0];
}

const PIPE_PATH       = '\\\\.\\pipe\\HmiRuntime';
const BROWSE_COOKIE   = 'openpipe-browse-1';
const SUB_COOKIE      = 'openpipe-sub-1';
const PRE_SEND_DELAY_MS = 1000;
const PROBE_WAIT_MS     = 5000;

log('Pipe path:  ' + PIPE_PATH);
log('Filter:     ' + filter);
log('Skip probe: ' + skipProbe);

let rxCount = 0;
let lastRxTime = Date.now();
let probeAnswered = false;
const browsedTags = [];   // accumulates tag names across pages

const client = net.createConnection(PIPE_PATH);

client.on('connect', () => {
    log('Pipe connected');

    const rl = readline.createInterface({ input: client, crlfDelay: Infinity });
    rl.on('line', (line) => {
        lastRxTime = Date.now();
        rxCount++;
        log('RX #' + rxCount + ': ' + line);
        probeAnswered = true;

        try {
            const msg = JSON.parse(line);
            switch (msg.Message) {

                case 'NotifyBrowseTags': {
                    const tags = (msg.Params && msg.Params.Tags) || [];
                    if (tags.length === 0) {
                        log('BrowseTags complete (empty page) — ' + browsedTags.length + ' tags total');
                        sendSubscribe(browsedTags);
                    } else {
                        tags.forEach(t => browsedTags.push(t.Name));
                        log('  -> page of ' + tags.length + ' tags (total so far: ' + browsedTags.length + ')');
                        sendNext();
                    }
                    break;
                }

                case 'ErrorBrowseTags': {
                    // Server closed the browse session (last partial page already sent).
                    log('ErrorBrowseTags (' + msg.ErrorDescription + ') — treating as browse complete, ' + browsedTags.length + ' tags');
                    sendSubscribe(browsedTags);
                    break;
                }

                case 'NotifySubscribeTag': {
                    const tags = (msg.Params && msg.Params.Tags) || [];
                    const changed   = tags.filter(t => t.hasChanged !== false);
                    const unchanged = tags.filter(t => t.hasChanged === false);
                    log('  -> ' + tags.length + ' total: ' + changed.length + ' changed, ' + unchanged.length + ' unchanged (hasChanged=false)');
                    changed.forEach(t => log('     CHANGED ' + t.Name + ' = ' + t.Value + ' [' + t.Quality + ']'));
                    break;
                }
            }
        } catch (e) {
            log('  (parse error: ' + e.message + ')');
        }
    });

    log('Waiting ' + PRE_SEND_DELAY_MS + 'ms before first send...');
    setTimeout(() => {
        if (skipProbe) {
            sendBrowse();
            return;
        }

        sendLine('PROBE', JSON.stringify({
            Message: 'ReadConfig',
            Params:  ['DefaultPageSize'],
            ClientCookie: 'openpipe-test-probe'
        }));

        const deadline = Date.now() + PROBE_WAIT_MS;
        const waitForProbe = () => {
            if (probeAnswered) {
                log('Probe round-trip OK');
                setTimeout(sendBrowse, 500);
            } else if (Date.now() < deadline) {
                setTimeout(waitForProbe, 100);
            } else {
                log('WARNING: no response to probe after ' + PROBE_WAIT_MS + 'ms');
                log('RT is not reading our bytes. Try restarting WinCC Unified Runtime.');
                setTimeout(sendBrowse, 500);
            }
        };
        setTimeout(waitForProbe, 100);
    }, PRE_SEND_DELAY_MS);
});

client.on('error', (err) => {
    log('Pipe error: ' + err.message);
    if (err.code === 'ENOENT') log('Pipe does not exist — is WinCC Unified Runtime running?');
    else if (err.code === 'EACCES') log('Access denied — is the current user in the SIMATIC HMI group?');
});

client.on('close', () => { log('Pipe closed'); });

function sendBrowse() {
    sendLine('BROWSE', JSON.stringify({
        Message: 'BrowseTags',
        Params:  { Filter: filter },
        ClientCookie: BROWSE_COOKIE
    }));
}

function sendNext() {
    sendLine('NEXT', JSON.stringify({
        Message: 'BrowseTags',
        Params:  'Next',
        ClientCookie: BROWSE_COOKIE
    }));
}

function sendSubscribe(tags) {
    if (tags.length === 0) { log('No tags to subscribe to.'); return; }
    log('Subscribing to ' + tags.length + ' tags...');
    sendLine('SUBSCRIBE', JSON.stringify({
        Message: 'SubscribeTag',
        Params:  { Tags: tags },
        ClientCookie: SUB_COOKIE
    }));
}

function sendLine(label, json) {
    log(label + ' TX: ' + json);
    client.write(json + '\n', 'utf8', () => { log(label + ' TX flushed'); });
}

function log(msg) {
    console.log('[' + new Date().toISOString() + '] ' + msg);
}

// Heartbeat every 5 s
setInterval(() => {
    const sinceRx = Math.round((Date.now() - lastRxTime) / 1000);
    log('heartbeat — RX count=' + rxCount + ', ' + sinceRx + 's since last RX');
}, 5000).unref();

File: dev/OpenPipeTest.java
// WinCC Unified Open Pipe — minimal standalone test program.
//
// Strips away Vert.X and the broker entirely. Confirms whether the round-trip with the
// local WinCC Unified Runtime over the named pipe works at the JDK level.
//
// Requirements:
//   - WinCC Unified Runtime running on this host
//   - The user running this command is a member of group "SIMATIC HMI" (Windows) or
//     "industrial" (Linux)
//   - JDK 11+ for single-file source launch (no compile step)
//
// Run from the repo root:
//   java dev\OpenPipeTest.java                 (default filter "*", probe first)
//   java dev\OpenPipeTest.java "HMI_Tag_*"     (custom filter)
//   java dev\OpenPipeTest.java skip "HMI_Tag_*"  (skip probe, go straight to BrowseTags)
//
// The program:
//   1. Opens \\.\pipe\HmiRuntime (Windows) or /tmp/HmiRuntime (Linux)
//   2. Starts a reader thread that prints every line received (with a timestamp)
//   3. Waits for the reader thread to signal it is running (CountDownLatch)
//   4. Waits an additional 3 s (READER_READY_WAIT_MS + PRE_SEND_DELAY_MS) for RT to attach
//      NOTE: if "TX flushed" never appears, RT isn't reading — restart WinCC RT and retry.
//      Windows named pipes are full-duplex: reads and writes are independent kernel operations,
//      so the reader thread and the main-thread writes run in parallel safely.
//   5. Sends a tiny `ReadConfig DefaultPageSize` probe first — simplest command, fast round-trip.
//      Waits up to 5 s for a response before proceeding.
//   6. Sends one BrowseTags request (expert/JSON syntax)
//   7. Sends Next requests in a loop until the server returns an empty page
//   8. Keeps running until Ctrl+C, printing a heartbeat every 5s.

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.atomic.AtomicLong;

public class OpenPipeTest {

    private static final String COOKIE = "openpipe-test-1";
    // Windows named pipes are full-duplex: the kernel keeps separate read/write buffers, so
    // ReadFile and WriteFile on the same handle can run concurrently from different threads.
    // RandomAccessFile("rw") opens with GENERIC_READ|GENERIC_WRITE; wrapping its FD with both
    // FileInputStream and FileOutputStream is safe — each issues an independent syscall.
    private static final long READER_READY_WAIT_MS = 2_000L;  // wait after reader thread signals ready
    private static final long PRE_SEND_DELAY_MS = 1_000L;     // extra gap before first send
    private static final long PROBE_WAIT_MS = 5_000L;         // wait this long for probe response
    private static final long INTER_STEP_DELAY_MS = 500L;     // small gap between commands
    private static BufferedWriter sharedWriter;               // synchronized below

    public static void main(String[] args) throws Exception {
        boolean skipProbe = false;
        String filter = "*";
        if (args.length > 0 && "skip".equalsIgnoreCase(args[0])) {
            skipProbe = true;
            if (args.length > 1) filter = args[1];
        } else if (args.length > 0) {
            filter = args[0];
        }

        String pipePath = System.getProperty("os.name", "").toLowerCase().contains("win")
            ? "\\\\.\\pipe\\HmiRuntime"
            : "/tmp/HmiRuntime";

        log("Pipe path:  " + pipePath);
        log("Filter:     " + filter);
        log("Skip probe: " + skipProbe);

        RandomAccessFile raf;
        try {
            raf = new RandomAccessFile(pipePath, "rw");
        } catch (FileNotFoundException e) {
            log("Cannot open pipe: " + e.getMessage());
            log("Is WinCC Unified Runtime running and is the current user in the SIMATIC HMI group?");
            System.exit(1);
            return;
        }
        log("Pipe opened");

        BufferedReader reader = new BufferedReader(
            new InputStreamReader(new FileInputStream(raf.getFD()), StandardCharsets.UTF_8));
        sharedWriter = new BufferedWriter(
            new OutputStreamWriter(new FileOutputStream(raf.getFD()), StandardCharsets.UTF_8));

        AtomicLong lastRx = new AtomicLong(System.currentTimeMillis());
        AtomicLong rxCount = new AtomicLong();
        CountDownLatch readerReady = new CountDownLatch(1);

        Thread readerThread = new Thread(() -> {
            log("Reader thread running — signalling ready");
            readerReady.countDown();   // signal BEFORE blocking on readLine
            try {
                String line;
                while ((line = reader.readLine()) != null) {
                    lastRx.set(System.currentTimeMillis());
                    long n = rxCount.incrementAndGet();
                    log("RX #" + n + ": " + line);
                    if (line.contains("\"Message\":\"NotifyBrowseTags\"")) {
                        boolean emptyPage = line.contains("\"Tags\":[]") || !line.contains("\"Tags\":[");
                        if (emptyPage) {
                            log("BrowseTags complete (empty/no-Tags page)");
                        } else {
                            sendNext();
                        }
                    }
                }
                log("Reader: EOF");
            } catch (IOException e) {
                log("Reader IOException: " + e.getMessage());
            }
        }, "openpipe-test-reader");
        readerThread.setDaemon(true);
        readerThread.start();

        log("Waiting for reader thread ready signal...");
        readerReady.await();
        log("Reader thread ready. Waiting " + READER_READY_WAIT_MS + "ms before first send...");
        Thread.sleep(READER_READY_WAIT_MS);
        log("Pre-send delay done. Waiting additional " + PRE_SEND_DELAY_MS + "ms...");
        Thread.sleep(PRE_SEND_DELAY_MS);

        if (!skipProbe) {
            // Probe: read one config value. This is the simplest command in the protocol.
            // If we don't see "TX flushed" within a few seconds, the server isn't reading
            // our bytes at all — that's an RT-side state issue, restart WinCC RT.
            String probe = "{\"Message\":\"ReadConfig\","
                         + "\"Params\":[\"DefaultPageSize\"],"
                         + "\"ClientCookie\":\"openpipe-test-probe\"}";
            sendLine("PROBE", probe);
            log("Waiting up to " + PROBE_WAIT_MS + "ms for probe response...");
            long deadline = System.currentTimeMillis() + PROBE_WAIT_MS;
            long initialRxCount = rxCount.get();
            while (rxCount.get() == initialRxCount && System.currentTimeMillis() < deadline) {
                Thread.sleep(100);
            }
            if (rxCount.get() == initialRxCount) {
                log("WARNING: no response to probe after " + PROBE_WAIT_MS + "ms");
                log("This usually means WinCC Unified Runtime isn't reading the pipe.");
                log("Try restarting WinCC Unified Runtime and re-running this test.");
            } else {
                log("Probe round-trip OK");
            }
            Thread.sleep(INTER_STEP_DELAY_MS);
        }

        String browseRequest = "{\"Message\":\"BrowseTags\","
                             + "\"Params\":{\"Filter\":\"" + escape(filter) + "\"},"
                             + "\"ClientCookie\":\"" + COOKIE + "\"}";
        sendLine("BROWSE", browseRequest);

        // Heartbeat loop — keep running until Ctrl+C.
        long lastHeartbeat = System.currentTimeMillis();
        while (true) {
            Thread.sleep(1_000);
            long now = System.currentTimeMillis();
            if (now - lastHeartbeat >= 5_000) {
                long sinceRx = now - lastRx.get();
                log("heartbeat — RX count=" + rxCount.get() + ", " + (sinceRx / 1000) + "s since last RX");
                lastHeartbeat = now;
            }
        }
    }

    private static void sendNext() {
        String next = "{\"Message\":\"BrowseTags\",\"Params\":\"Next\",\"ClientCookie\":\"" + COOKIE + "\"}";
        sendLine("NEXT", next);
    }

    private static void sendLine(String label, String line) {
        try {
            log(label + " TX: " + line);
            synchronized (sharedWriter) {
                sharedWriter.write(line);
                sharedWriter.write("\n");
                sharedWriter.flush();
            }
            log(label + " TX flushed");
        } catch (IOException e) {
            log(label + " write failed: " + e.getMessage());
        }
    }

    private static String escape(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private static void log(String msg) {
        System.out.println("[" + Instant.now() + "] " + msg);
    }
}
